host_provider/providers/azure.py

```diff
@@ -0,0 +1,342 @@
+        self.get_azure_connection()
+        self.azClient.connect(base_url=base_url)
+        self.azClient.add_default_headers(header)
+        self.azClient.connection.request("GET", action, headers=header)
+        resp = self.azClient.connection.getresponse()
+
+        if resp.status_code == 200:
+            offerings = resp.json()['value']
+
+            for offering in offerings:
+                if offering.get('memoryInMB') == memory and offering.get('numberOfCores') == cpu:
+                    return offering
+
+        raise OfferingNotFoundError(
+            "Offering with {} cpu and {} of memory not found.".format(cpu, memory)
+        )
+
+    def _parse_nic(self, name, vnet, subnet, version='1.0.0'):
+        nic_dict = OrderedDict()
+        id = '/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Network/virtualNetworks/%s/subnets/%s' \
+             % (self.credential.subscription_id, self.credential.resource_group, vnet, subnet)
+
+        region = self.credential.region
+        templates = JsonTemplates()
+        try:
+            for file in templates.list_files(version):
+                file_name = file.name.split('.json')[0]
+                if file_name == 'nic':
+                    nic_dict = templates.load_json(file.as_posix())
+                    config = nic_dict['properties']['ipConfigurations']
+                    for nic in config:
+                        nic['name'] = name
+                        nic['properties']['subnet']['id'] = id
+                    nic_dict['properties']['ipConfigurations'] = config
+                    nic_dict['location'] = region
+            return nic_dict
+        except Exception as error:
+            raise Exception("Template not found error")
+
+    def create_nic(self, name, api_version='2020-07-01'):
+        subnet = self.credential.get_next_zone_from(self.credential.subnets)
+        vnet = self.credential.subnets.get(subnet)['name']
+        base_url = self.credential.endpoint
+        action = 'subscriptions/%s/resourceGroups/%s/providers/Microsoft.Network/networkInterfaces/%s?api-version=%s' \
+                 % (self.credential.subscription_id, self.credential.resource_group, name, api_version)
+        nic = self._parse_nic(name, vnet, subnet)
+
+        payload = json.dumps(nic)
+        header = {}
+        self.get_azure_connection()
+        self.azClient.connect(base_url=base_url)
+        self.azClient.add_default_headers(header)
+        self.azClient.connection.request("PUT", action, body=payload, headers=header)
+        resp = self.azClient.connection.getresponse()
+
+        return resp
+
+    def get_network(self, api_version='2020-07-01'):
+        subnet = self.credential.get_next_zone_from(self.credential.subnets)
+        vnet = self.credential.subnets.get(subnet)['name']
+        base_url = self.credential.endpoint
+
+        action = "subscriptions/%s/resourceGroups/%s/providers/Microsoft.Network/virtualNetworks/%s?api-version=%s" \
```

Comment:
- Hardcoded url.
- use format instead of %
- do no use \ to break lines, use parentesis